pytng/pytng.pyx

```diff
@@ -93,25 +407,206 @@ cdef class MemoryWrapper:
+        elif self.mode == 'w':
+            _mode = 'w'
+            raise NotImplementedError('Writing is not implemented yet.')
+        elif self.mode == 'a':
+            _mode = 'a'
+            raise NotImplementedError('Appending is not implemented yet')
+        else:
+            raise ValueError('mode must be one of "r", "w", or "a" you '
+                             'supplied {}'.format(mode))
+
+        # handle file not existing at python level,
+        # C level is nasty and causes crash
+        if self.mode == 'r' and not os.path.isfile(fname):
+            raise IOError("File '{}' does not exist".format(fname))
+
+        cdef tng_function_status stat
+
+        fname_bytes = fname.encode('UTF-8')
+        stat = tng_util_trajectory_open(fname_bytes, _mode, & self._traj)
+        if stat != TNG_SUCCESS:
+            raise IOError("File '{}' cannot be opened".format(fname))
+
+        # TODO propagate errmsg upwards in all the below calls
+        stat = tng_num_frames_get(self._traj, & self._n_frames)
+        if stat != TNG_SUCCESS:
+            raise IOError("Number of frames cannot be read")
+
+        stat = tng_num_particles_get(self._traj, & self._n_particles)
+        if stat != TNG_SUCCESS:
+            raise IOError("Number of particles cannot be read")
+
+        # NOTE can we just loop over this directly in some way?
+        stat = tng_num_frame_sets_get(self._traj, & self._n_frame_sets)
+        # TODO can this be read straight from the struct as self._traj->n_frame_sets?
+        # they note that this is not always updated
+
+        cdef int64_t exponent
+        stat = tng_distance_unit_exponential_get(self._traj, & exponent)
+        if stat != TNG_SUCCESS:
+            raise IOError("Distance exponent cannot be read")
+
+        self._distance_scale = 10.0**(exponent+9)
+
+        self.is_open = True
+        self.step = 0
+        self.reached_eof = False
+    
+
+    def _close(self):
+        """Make sure the file handle is closed"""
+        if self.is_open:
+            tng_util_trajectory_close(& self._traj)
+            self.is_open = False
+            self._n_frames = -1
+
+    def spool(self):
+        cdef tng_function_status stat = TNG_SUCCESS
+        cdef int64_t block_count = 0
+        while stat != TNG_CRITICAL:
+            print("block read called {}".format(block_count))
+            stat = self._read_next_block()
+            block_count += 1
+
+    # NOTE this looks simple and may work ? may be too low level as it all hinges on whether the file positions are kept up to date
+    # can we just fseek to the first TFS?
+    cdef tng_function_status _read_next_block(self):
+        cdef tng_function_status stat
+        cdef tng_gen_block* block
+        cdef int64_t block_id = -1
+        stat = tng_block_init(& block)
+        if stat != TNG_SUCCESS:
+            return TNG_CRITICAL
+        
+        fseeko(dereference(self._traj).input_file, 0, SEEK_SET) # set to start of inut file
```

Comment:
I think you can just do `self._traj.input_file` and Cython knows you mean `self._traj->input_file`

Reply:
Okay i'll change :)